### PR TITLE
[RFC] MIPS Base Refine

### PR DESCRIPTION
**Pull request by FlyGoat**
## 拉取/合并请求描述：(PR description)

[
这个PR主要是对MIPS基础代码进行重构，并且添加QEMU mipssim bsp支持，并且目前已经可以在QEMU中运行。
目前仅仅完成了common代码的重构，其他各个BSP和cpuport还需要进一步修改。
我会完成龙芯1B/1C和君正X1000的适配。龙芯部分我会自行测试，君正部分还需要麻烦 @Xburst 合作。
Request For Comments from reviewers. 
]

以下的内容不应该在提交PR时的message修改，修改下述message，PR会被直接关闭。请在提交PR后，浏览器查看PR并对以下检查项逐项check，没问题后逐条在页面上打钩。
The following content must not be changed in submitted PR message. Otherwise, the PR will be closed immediately. After submitted PR, please use web browser to visit PR, and check items one by one, and ticked them if no problem.

### 当前拉取/合并请求的状态 Intent for your PR

必须选择一项 Choose one (Mandatory):

- [x] 本拉取/合并请求是一个草稿版本 This PR is for a code-review and is intended to get feedback
- [ ] 本拉取/合并请求是一个成熟版本 This PR is mature, and ready to be integrated into the repo

### 代码质量 Code Quality：

我在这个拉取/合并请求中已经考虑了 As part of this pull request, I've considered the following:

- [ ] 已经仔细查看过代码改动的对比 Already check the difference between PR and old code
- [ ] 代码风格正确，包括缩进空格，命名及其他风格 Style guide is adhered to, including spacing, naming and other style
- [ ] 没有垃圾代码，代码尽量精简，不包含`#if 0`代码，不包含已经被注释了的代码 All redundant code is removed and cleaned up
- [ ] 所有变更均有原因及合理的，并且不会影响到其他软件组件代码或BSP All modifications are justified and not affect other components or BSP
- [ ] 对难懂代码均提供对应的注释 I've commented appropriately where code is tricky
- [ ] 本拉取/合并请求代码是高质量的 Code in this PR is of high quality

### CODE REVIEW

**Comment by FlyGoat** [comment]
BTW: It seems like there is no actual BSP is using libcpu/xburst, could we simply drop it?
For PIC32, we'd better place the libcpu into BSP as it's not very generic.
Refined code requires MIPS32R2+ ISA as it's using Ebase feature, and ehb to deal with hazard issue.

File: bsp/mipssim/applications/SConscript [hunk 1]
```diff
@@ -0,0 +1,9 @@
+from building import *
+
+cwd     = GetCurrentDir()
+src	= Glob('*.c')
+CPPPATH = [cwd, str(Dir('#'))]
```
Comment (BernardXiong) [inline bsp/mipssim/applications/SConscript]: Please take care of indentation. And no need to add `str(Dir('#'))`

File: bsp/mipssim/drivers/board.c [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+/*@{*/
+
+#include <rtthread.h>
+#include <rthw.h>
+
+#include "mips_regs.h"
+#include "exception.h"
+#include "drv_uart.h"
+
+#define CPU_HZ	(100 * 1000 * 1000)
```
Comment (BernardXiong) [inline bsp/mipssim/drivers/board.c]: Please take care of indentation.

File: bsp/mipssim/drivers/board.c [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    write_c0_count(0);
+    mips_unmask_cpu_irq(7);
+}
+
+
+/**
+ * init hardware FPU
+ */
+#ifdef RT_USING_FPU
+void rt_hw_fpu_init(void)
```
Comment (BernardXiong) [inline bsp/mipssim/drivers/board.c]: Whether FPU working?

File: bsp/mipssim/drivers/drv_uart.c [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+    /* device initialization */
+    uart->parent.type = RT_Device_Class_Char;
+    rt_memset(uart->rx_buffer, 0, sizeof(uart->rx_buffer));
+    uart->read_index = uart->save_index = 0;
+
+    uart->hw_base = UART0_BASE;
+    uart->irq = 4;
+
+    /* device interface */
+    uart->parent.init       = rt_uart_init;
```
Comment (BernardXiong) [inline bsp/mipssim/drivers/drv_uart.c]: Please use serial device driver, and implement putc/getc.

File: bsp/mipssim/drivers/drv_uart.h [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+#define UART_MCR(base)		__REG8(base + 0x04)
+#define UART_LSR(base)		__REG8(base + 0x05)
+#define UART_MSR(base)		__REG8(base + 0x06)
+
+#define UART_LSB(base)		__REG8(base + 0x00)
+#define UART_MSB(base)		__REG8(base + 0x01)
+
+
+/* UART interrupt enable register value */
+#define UARTIER_IME		(1 << 3)
```
Comment (BernardXiong) [inline bsp/mipssim/drivers/drv_uart.h]: Please take care of indentation.

File: libcpu/mips/common/exc_gcc.S [hunk 6]
```diff
@@ -0,0 +1,59 @@
+/*
```
Comment (BernardXiong) [inline libcpu/mips/common/exc_gcc.S]: better with name as exception_gcc.S

File: libcpu/mips/common/exception.c [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+    for (i=0; i<RT_EXCEPTION_MAX; i++)
+        sys_exception_handlers[i] = (exception_func_t)unhandled_exception_handle;
+}
+
+int rt_hw_exception_init(void)
+{
+    rt_uint32_t ebase = (rt_uint32_t)&__ebase_entry;
+    write_c0_ebase(ebase);
+    rt_kprintf("ebase: %lx\n",ebase);
```
Comment (BernardXiong) [inline libcpu/mips/common/exception.c]: better with LOG_D.

File: libcpu/mips/common/exception.c [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+    return RT_EOK;
+}
+
+void rt_general_exc_dispatch(struct pt_regs *regs)
+{
+    rt_uint32_t cause = read_c0_cause();
+    rt_uint32_t index;
+    rt_uint32_t exccode = (cause & CAUSEF_EXCCODE) >> CAUSEB_EXCCODE;
+
+    if (exccode == 0) {
```
Comment (BernardXiong) [inline libcpu/mips/common/exception.c]: Please take care of code style.

File: libcpu/mips/common/stack.c [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+#include <rtthread.h>
+
+#include "mips.h"
+
+register rt_uint32_t $GP __asm__ ("$28");
+
+rt_uint8_t *rt_hw_stack_init(void *tentry, void *parameter, rt_uint8_t *stack_addr, void *texit)
+{
+	static rt_uint32_t wSR=0;
```
Comment (BernardXiong) [inline libcpu/mips/common/stack.c]: Please use space for indenation.

File: bsp/mipssim/drivers/board.c (hunk 3 above)
Comment (FlyGoat) [reply bsp/mipssim/drivers/board.c]: Tests in QEMU suggest it works. But I'm not sure if it's perfect. So I'd like to disable it deafultly.

**Comment by FlyGoat** [comment]
Thanks for your kindly review~ Sorry for the styling issues. I get used to with Linux Kernel Coding style and It's really pain for me to switch between them:-) Next revision will surely deal these issues.
Do we have any tool like 'checkpatch.pl' in Linux to check it automatically?

**Comment by BernardXiong** [comment]
> Thanks for your kindly review~ Sorry for the styling issues. I get used to with Linux Kernel Coding style and It's really pain for me to switch between them:-) Next revision will surely deal these issues.
> Do we have any tool like 'checkpatch.pl' in Linux to check it automatically?

No checking tool yet so far.

**Comment by BernardXiong** [comment]
@XBurst Please verify X1000 at same time or be removed from RT-Thread BSP. Thanks.

**Comment by BernardXiong** [comment]
@FlyGoat 
Please note that the ls1bdev/ls1cdev and X1000 BSP travis-ci failed.

**Comment by FlyGoat** [comment]
@BernardXiong I have resolved your review suggestions and pushed ls1b/ls1c/x1000 code. Please review, thanks.

@Xburst Please help with testing on X1000, prebuilt binary attached, thanks.
[rtthread-x1000-exp.zip](https://github.com/RT-Thread/rt-thread/files/3933242/rtthread-x1000-exp.zip)

**Comment by FlyGoat** [comment]
TODO: Further Debug support.
Experiments revealed that origin backtrace implementation in Xburst BSP is somehow broken. Backtrace/register dump/exception handling is necessary in debugging. I'm going to work on that after this PR being merged.

File: bsp/ls1bdev/drivers/uart.h [hunk 10]
```diff
@@ -31,43 +31,43 @@
...
 /* UART registers */
-#define UART_DAT(base)		__REG8(base + 0x00)
-#define UART_IER(base)		__REG8(base + 0x01)
-#define UART_IIR(base)		__REG8(base + 0x02)
-#define UART_FCR(base)		__REG8(base + 0x02)
-#define UART_LCR(base)		__REG8(base + 0x03)
-#define UART_MCR(base)		__REG8(base + 0x04)
-#define UART_LSR(base)		__REG8(base + 0x05)
-#define UART_MSR(base)		__REG8(base + 0x06)
+#define UART_DAT(base)		REG8(base + 0x00)
```
Comment (BernardXiong) [inline bsp/ls1bdev/drivers/uart.h]: You can use `rthw.h/HWREG8` to replace REG8.

File: bsp/ls1cdev/SConstruct [hunk 11]
```diff
@@ -29,7 +29,7 @@ Export('rtconfig')
 objs = PrepareBuilding(env, RTT_ROOT)
 
 if GetDepend('RT_USING_SELF_BOOT'):
-    rtconfig.LFLAGS += " -Ttext 0x80010000"
+    rtconfig.LFLAGS += " -Ttext 0x80000000"
 else:
 	rtconfig.LFLAGS += " -Ttext 0x80200000"
```
Comment (BernardXiong) [inline bsp/ls1cdev/SConstruct]: Could you please cleanup the indention (space/tab) inside this file? Thanks.

File: libcpu/mips/common/ptrace.h [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+	/* FPU Registers */
+	/* Unlike Linux Kernel, we save these registers unconditionally,
+	 * so it should be a part of pt_regs */
+	struct mips_fpu_struct fpu;
+#endif
+} __aligned(8);
+#endif
+
+/* Note: For call stack o32 ABI has 0x8 shadowsoace Here  */
+#define PT_R0		(0x8 * LONGSIZE)	/* 0 */
```
Comment (BernardXiong) [inline libcpu/mips/common/ptrace.h]: Please take care of indention.

File: libcpu/mips/gs232/interrupt.c [hunk 13]
```diff
@@ -0,0 +1,164 @@
+/*
+ * File      : interrupt.c
+ * This file is part of RT-Thread RTOS
+ * COPYRIGHT (C) 2006 - 2011, RT-Thread Development Team
+ *
+ * The license and distribution terms for this file may be
+ * found in the file LICENSE in this distribution or at
+ * http://www.rt-thread.org/license/LICENSE
```
Comment (BernardXiong) [inline libcpu/mips/gs232/interrupt.c]: Please use correct file header.

File: libcpu/mips/gs232/ls1b.h [hunk 14]
```diff
@@ -0,0 +1,71 @@
+/*
+ * File      : ls1b.h
+ * This file is part of RT-Thread RTOS
+ * COPYRIGHT (C) 2006-2011, RT-Thread Develop Team
+ *
+ * The license and distribution terms for this file may be
+ * found in the file LICENSE in this distribution or at
+ * http://www.rt-thread.org/license/LICENSE
```
Comment (BernardXiong) [inline libcpu/mips/gs232/ls1b.h]: Please use correct file header.

**Comment by FlyGoat** [comment]
@BernardXiong Done, Thanks for your review :+1:

**Comment by BernardXiong** [comment]
@FlyGoat I move the libcpu/mips/x1000 to bsp/x1000. Please re-submit your PR, thanks.

**Pull request closed without merging**